rust/kernel/sync/spinlock.rs

```diff
@@ -178,3 +178,126 @@ unsafe impl<T: ?Sized> Lock for SpinLock<T> {
+/// fn lock_irqdisable_example(value: &IrqDisableSpinLock<Example>) {
+///     let mut guard = value.lock();
+///     guard.a = 10;
+///     guard.b = 20;
+/// }
+///
+/// // Initialises a spinlock and calls the example functions.
+/// pub fn spinlock_example() {
+///     // SAFETY: `spinlock_init` is called below.
+///     let mut value = unsafe { IrqDisableSpinLock::new(Example { a: 1, b: 2 }) };
+///     // SAFETY: We don't move `value`.
+///     kernel::irqdisable_spinlock_init!(unsafe { Pin::new_unchecked(&mut value) }, "value");
+///     lock_irqdisable_example(&value);
+/// }
+/// ```
+pub struct IrqDisableSpinLock<T: ?Sized> {
+    inner: SpinLock<T>,
+}
+
+// SAFETY: `IrqDisableSpinLock` can be transferred across thread boundaries iff the data it protects
+// can.
+unsafe impl<T: ?Sized + Send> Send for IrqDisableSpinLock<T> {}
+
+// SAFETY: `IrqDisableSpinLock` serialises the interior mutability it provides, so it is `Sync` as
+// long as the data it protects is `Send`.
+unsafe impl<T: ?Sized + Send> Sync for IrqDisableSpinLock<T> {}
+
+impl<T> IrqDisableSpinLock<T> {
+    /// Constructs a new irq-disabling spinlock.
+    ///
+    /// # Safety
+    ///
+    /// The caller must call [`IrqDisableSpinLock::init_lock`] before using the spinlock.
+    pub unsafe fn new(t: T) -> Self {
+        // SAFETY: By the safety requirements of this function, The caller is required to call
+        // `IrqDisableSpinLock::init_lock`, which itself calls `inner.init_lock`.
+        Self {
+            inner: unsafe { SpinLock::new(t) },
+        }
+    }
+}
+
+impl<T: ?Sized> IrqDisableSpinLock<T> {
+    /// Locks the spinlock and gives the caller access to the data protected by it. Additionally it
+    /// disables interrupts (if they are enabled).
+    ///
+    /// When the lock in unlocked, the interrupt state (enabled/disabled) is restored.
+    pub fn lock(&self) -> GuardMut<'_, Self> {
+        let ctx = self.lock_noguard();
+        // SAFETY: The spinlock was just acquired.
+        unsafe { GuardMut::new(self, ctx) }
+    }
+}
+
+impl<T> CreatableLock for IrqDisableSpinLock<T> {
+    unsafe fn new_lock(data: Self::Inner) -> Self {
+        // SAFETY: The safety requirements of `new_lock` also require that `init_lock` be called.
+        unsafe { Self::new(data) }
+    }
+
+    unsafe fn init_lock(
+        mut self: Pin<&mut Self>,
+        name: &'static CStr,
+        key: *mut bindings::lock_class_key,
+    ) {
+        // SAFETY: `self.inner` is pinned when `self` is.
+        let inner = unsafe { self.as_mut().map_unchecked_mut(|s| &mut s.inner) };
+        // SAFETY: By the safety requirements of this function, `key` will remain valid.
+        unsafe { inner.init_lock(name, key) };
+    }
+}
+
+// SAFETY: The underlying kernel `spinlock_t` object ensures mutual exclusion.
+unsafe impl<T: ?Sized> Lock for IrqDisableSpinLock<T> {
+    type Inner = T;
+    type GuardContext = <SpinLock<T> as Lock>::GuardContext;
```

Comment:
I think this can be
```rust
type GuardContext = ();
```
and `lock_noguard()` is `spin_lock_irq()`, `unlock()` is `spin_unlock_irq()`. RIght?

Reply:
> I think this can be
> 
> ```rust
> type GuardContext = ();
> ```
> 
> and `lock_noguard()` is `spin_lock_irq()`, `unlock()` is `spin_unlock_irq()`. RIght?
> 
> Of course, if this can help compiler generate better code.

No, it needs to be `spin_lock_irqsave` and `spin_unlock_irqrestore`. I'll repeat the description above here:

> It is a spinlock that always disables interrupts when locked. It is a
> convenience type that makes it harder for users to forget to disable
> interrupts when acquiring a spinlock that needs interrupts disabled.

IOW, if interrupts are already disabled, they remain disabled when we call `lock`, and also when we call `unlock`. If we switch to `spin_unlock_irq`, we will re-enable interrupts on `unlock`, which we don't want in this case. (We only want to re-enable interrupts on `unlock` if they were enabled when `lock` was called, which is the semantics offered by irqsave/irqrestore.)

Reply:
Ok, I failed to notice this, sorry ;-(

However, while we are at it, I wonder whether `IrqDisableSpinLock::relock()` breaks this invariant. Because users can call `some_irqdisable_spinlock.relock(&mut None)` which will eventually call `SpinLock::relock()`, and it doesn't disable the irq.

So instead of `Option<ulong>`, maybe we want to define `IrqDisableSpinLock::GuardContext` as `ulong`?